server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -932,83 +930,100 @@ public int deletePendingSegments(String dataSource)
+      final List<DataSegment> segmentList = new ArrayList<>(toInsertSegments);
+
+      PreparedBatch preparedBatch = handle.prepareBatch(
           StringUtils.format(
-              "INSERT INTO %1$s (id, dataSource, created_date, start, %2$send%2$s, partitioned, version, used, "
-              + "payload) "
-              + "VALUES (:id, :dataSource, :created_date, :start, :end, :partitioned, :version, :used, :payload)",
+              "INSERT INTO %1$s (id, dataSource, created_date, start, %2$send%2$s, partitioned, version, used, payload) "
+                  + "VALUES (:id, :dataSource, :created_date, :start, :end, :partitioned, :version, :used, :payload)",
               dbTables.getSegmentsTable(),
               connector.getQuoteString()
           )
-      )
+      );
+
+      for (int i = 0; i < segmentList.size(); i++) {
+        DataSegment segment = segmentList.get(i);
+        preparedBatch.add()
             .bind("id", segment.getId().toString())
             .bind("dataSource", segment.getDataSource())
             .bind("created_date", DateTimes.nowUtc().toString())
             .bind("start", segment.getInterval().getStart().toString())
             .bind("end", segment.getInterval().getEnd().toString())
             .bind("partitioned", (segment.getShardSpec() instanceof NoneShardSpec) ? false : true)
             .bind("version", segment.getVersion())
-            .bind("used", used)
-            .bind("payload", jsonMapper.writeValueAsBytes(segment))
-            .execute();
-
-      if (numRowsInserted == 1) {
-        log.info(
-            "Published segment [%s] to DB with used flag [%s], json[%s]",
-            segment.getId(),
-            used,
-            jsonMapper.writeValueAsString(segment)
-        );
-      } else if (numRowsInserted == 0) {
-        throw new ISE(
-            "Failed to publish segment[%s] to DB with used flag[%s], json[%s]",
-            segment.getId(),
-            used,
-            jsonMapper.writeValueAsString(segment)
-        );
-      } else {
-        throw new ISE(
-            "numRowsInserted[%s] is larger than 1 after inserting segment[%s] with used flag[%s], json[%s]",
-            numRowsInserted,
-            segment.getId(),
-            used,
-            jsonMapper.writeValueAsString(segment)
-        );
+            .bind("used", usedSegments.contains(segment))
+            .bind("payload", jsonMapper.writeValueAsBytes(segment));
+
+        if ((i + 1) % ANNOUNCE_HISTORICAL_SEGMENG_BATCH == 0 || i == segmentList.size() - 1) {
+          int[] affectedRows = preparedBatch.execute();
+          for (int j = 0; j < affectedRows.length; j++) {
+            DataSegment insertSegment = segmentList.get(i / ANNOUNCE_HISTORICAL_SEGMENG_BATCH * ANNOUNCE_HISTORICAL_SEGMENG_BATCH + j);
+            if (affectedRows[j] == 1) {
+              log.info(
+                  "Published segment [%s] to DB with used flag [%s], json[%s]",
```

Comment:
This will corrupt the logs too. How about modifying as the below?

```java
      final List<List<DataSegment>> partitionedSegments = Lists.partition(
          new ArrayList<>(toInsertSegments),
          MAX_NUM_SEGMENTS_TO_ANNOUNCE_AT_ONCE
      );

      PreparedBatch preparedBatch = handle.prepareBatch(
          StringUtils.format(
              "INSERT INTO %1$s (id, dataSource, created_date, start, %2$send%2$s, partitioned, version, used, payload) "
                  + "VALUES (:id, :dataSource, :created_date, :start, :end, :partitioned, :version, :used, :payload)",
              dbTables.getSegmentsTable(),
              connector.getQuoteString()
          )
      );
      
      for (List<DataSegment> partition : partitionedSegments) {
        for (DataSegment segment : partition) {
          preparedBatch.add()
                       .bind("id", segment.getId().toString())
                       .bind("dataSource", segment.getDataSource())
                       .bind("created_date", DateTimes.nowUtc().toString())
                       .bind("start", segment.getInterval().getStart().toString())
                       .bind("end", segment.getInterval().getEnd().toString())
                       .bind("partitioned", (segment.getShardSpec() instanceof NoneShardSpec) ? false : true)
                       .bind("version", segment.getVersion())
                       .bind("used", usedSegments.contains(segment))
                       .bind("payload", jsonMapper.writeValueAsBytes(segment));
        }
        final int[] affectedRows = preparedBatch.execute();
        final boolean succeeded = Arrays.stream(affectedRows).allMatch(eachAffectedRows -> eachAffectedRows == 1);
        if (succeeded) {
          log.infoSegments(partition, "Published segments to DB");
        } else {
          final List<DataSegment> failedToPublish = IntStream.range(0, partition.size())
                                                             .filter(i -> affectedRows[i] != 1)
                                                             .mapToObj(partition::get)
                                                             .collect(Collectors.toList());
          throw new ISE(
              "Failed to publish segments to DB: %s",
              SegmentUtils.commaSeparatedIdentifiers(failedToPublish)
          );
        }
      }
    }
```